client/blocks/subscription-length-picker/option.jsx

```diff
@@ -0,0 +1,126 @@
+	static propTypes = {
+		term: PropTypes.string.isRequired,
+		savePercent: PropTypes.number.isRequired,
+		price: PropTypes.string.isRequired,
+		pricePerMonth: PropTypes.string.isRequired,
+		checked: PropTypes.bool.isRequired,
+		value: PropTypes.any.isRequired,
+		onCheck: PropTypes.func.isRequired,
+		translate: PropTypes.func.isRequired,
+	};
+
+	static defaultProps = {
+		checked: false,
+		savePercent: 0,
+		onCheck: () => null,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.htmlId = uniqueId( 'subscription-option-' );
+	}
+
+	render() {
+		const { savePercent, price, term, checked } = this.props;
+		const className = classnames( 'subscription-length-picker__option', {
+			'is-active': checked,
+		} );
+		return (
+			<label className={ className } htmlFor={ this.htmlId }>
+				<div className="subscription-length-picker__option-radio-wrapper">
+					<input
+						id={ this.htmlId }
+						type="radio"
+						className="subscription-length-picker__option-radio"
+						checked={ checked }
+						onChange={ this.handleChange }
+					/>
+				</div>
+
+				<div className="subscription-length-picker__option-content">
+					<div className="subscription-length-picker__option-header">
+						<div className="subscription-length-picker__option-term">{ this.getTermText() }</div>
+						<div className="subscription-length-picker__option-discount">
+							{ savePercent ? this.renderSaveBadge() : false }
+						</div>
+					</div>
+					<div className="subscription-length-picker__option-description">
+						<div className="subscription-length-picker__option-price">{ price }</div>
+						<div className="subscription-length-picker__option-side-note">
+							{ term !== TERM_MONTHLY ? this.renderPricePerMonth() : false }
+						</div>
+					</div>
+				</div>
+			</label>
+		);
+	}
+
+	getTermText() {
+		const { term, translate } = this.props;
+		switch ( term ) {
+			case TERM_BIENNIALLY:
+				return translate( '%s year', '%s years', { count: 2, args: '2' } );
+
+			case TERM_ANNUALLY:
+				return translate( '%s year', '%s years', { count: 1, args: '1' } );
+
+			case TERM_MONTHLY:
+				return translate( '%s month', '%s months', { count: 1, args: '1' } );
+		}
+	}
+
+	renderSaveBadge() {
+		const { savePercent, checked, translate } = this.props;
+		return (
+			<Badge type={ checked ? 'success' : 'warning' }>
+				{ translate( 'Save %(percent)s%%', {
+					args: {
+						percent: savePercent,
+					},
+				} ) }
+			</Badge>
+		);
+	}
+
+	renderPricePerMonth() {
+		const { savePercent, pricePerMonth, translate } = this.props;
+
+		let retval;
+		if ( savePercent ) {
+			retval = translate( 'only %(price)s / month', { args: { price: pricePerMonth } } );
+		} else {
+			retval = translate( '%(price)s / month', { args: { price: pricePerMonth } } );
+		}
+		return retval;
+	}
+
+	handleChange = e => {
+		if ( e.target.checked ) {
+			this.props.onCheck( {
+				value: this.props.value,
```

Comment:
note that while this may not matter (I'm not really in a position to know because I don't understand the implications as much) we are introducing a concurrency bug by using the value of `this.props` inside the event handler. that is, we aren't guaranteed that `this.props.value` is still the value it was when the check was clicked. the props could change in the time between the click and the time this callback was executed.

we have ways to work around this.

1. pass data through the DOM
```js
<input
	data-price={ this.props.value }
	onClick={ this.props.onClick }
/>

handleChange = ( { target: { checked, dataset: { price } } ) => checked && this.props.onCheck( { value: price } );
```

something like that. obviously `this.props.onClick` is still a concurrency issue but it's probably more valid than the `value`

2. bind the change handler (discouraged)
```js
onClick={ this.handleChange( this.props.value ) }

handleChange = value => e => e.target.checked && this.props.onCheck( { value } )
```

3. do more complicated and unwarranted stuff tracking the value

Reply:
I think you overanalyze this a bit :-) This is a single-threaded environment where most actions take place in response to some type of user input. There are few stack frames between the click and the handler and (I think) it is unfolded synchronously. So any problematic situation would mean that some other update occured the exact millisecond when user clicked. Like: user would have to click something else in different part of the screen, move the mouse here in 0ms and without firing any JS events, and click on that. Also, React would have to prefer one over another (side note: it may work differently in another version of React)

Like, maybe this could happen in a highly theoretical environment, but practically it won't happen, and if it will then maybe old props should be used after all? Also, there are virtually no consequences - any additional complexity here is just not worth it. And even if it was, there is no good way around that:

1. If props changes, DOM will also be updated so we may get either value in the handler depending on the exact implementation of how React schedules things in current version.
2. Bind is discouraged as you noted - new function will be created at each render() which means `onClick` will receive new value after each `render()` and it has certain consequences.

Reply:
> I think you overanalyze this a bit :-)

My goal is not to complicate the code but rather to make sure that we are aware of the places where failures will appear in the app and that we all handle them if the risks warrant a response. In this case I am unaware of the risks but I recognize the vector through which subtle bugs will appear and I'm sharing that acknowledgement with you to make the informed decision.

> any problematic situation would mean that some other update occured the exact millisecond when user clicked…maybe this could happen in a highly theoretical environment, but practically it won't happen

it actually happens quite often on many popular websites. for instance, on my Github notifications page I regularly navigate back to the list of notifications, click on the one I want, and then it promptly takes me to an issue/PR I _didn't_ click on.

Javascript is single-threaded but that doesn't mean that the event loop can't get behind. Calypso has had more than its share of second-plus delays in responding to clicks, and in that second React might be processing and updating props.

> If props changes, DOM will also be updated so we may get either value in the handler depending on the exact implementation of how React schedules things in current version.

Actually the nice thing about passing data through the DOM is that we eliminate the concurrency bugs. At _least_ if it's a concurrency issue then the passed data will match what the customer saw when they clicked. You are right that there could still be an error, but it will be a well-specified error instead of a subtle one.

---

You can ignore this advice! Though in this case I think the way to eliminate the bug is easy enough there's little reason to not do it.

Reply:
> Actually the nice thing about passing data through the DOM is that we eliminate the concurrency bugs. At least if it's a concurrency issue then the passed data will match what the customer saw when they clicked.

I argue that the two are equivalent:
1. You click on the component. Independently a prop is updated, event handler is called only after the update and uses the older one.
2. You click on the component. Independently a prop is updated, change is propagated to data-price, event handler is called only after the update and since it extracts data from DOM it also gets the older value.

Since react has it's own event propagation system and doesn't give any guarantees about the exact timing of propagating changes to DOM, using data- wouldn't buy us anything. Github notifications don't use react but they are just case in point since what you click on is `<a href>`.

But really - I lead the development on a highly interactive real-time meetings app written in React and we had some timing problems, but this here just isn't a context where we should be worried about it. A similar case would be when the list is getting updated and you click on list item during that update, but then react reused that exact element you clicked to show another option. Or removed the previous one and replaced it with something else 1ms before your finger finished clicking. Most of the time there is nothing we can do aside of writing code that doesn't block JS thread for an extended periods of time.